src/main/java/org/jenkinsci/plugins/workflow/log/BufferedBuildListener.java

```diff
@@ -0,0 +1,81 @@
+ * The MIT License
+ *
+ * Copyright 2018 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.log;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import hudson.CloseProofOutputStream;
+import hudson.model.BuildListener;
+import hudson.remoting.RemoteOutputStream;
+import hudson.util.StreamTaskListener;
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.PrintStream;
+import org.jenkinsci.remoting.SerializableOnlyOverRemoting;
+
+/**
+ * Unlike {@link StreamTaskListener} this does not set {@code autoflush} on the reconstructed {@link PrintStream}.
+ * It also wraps on the remote side in {@link DelayBufferedOutputStream}.
+ */
+final class BufferedBuildListener implements BuildListener, Closeable, SerializableOnlyOverRemoting {
+
+    private final OutputStream out;
+    @SuppressFBWarnings(value = "SE_BAD_FIELD", justification = "using Replacement anyway, fields here are irrelevant")
+    private final PrintStream ps;
+
+    BufferedBuildListener(OutputStream out) throws IOException {
+        this.out = out;
+        ps = new PrintStream(out, false, "UTF-8");
+    }
+    
+    @Override public PrintStream getLogger() {
+        return ps;
+    }
+    
+    @Override public void close() throws IOException {
+        ps.close();
+    }
+
+    private Object writeReplace() {
+        return new Replacement(this);
+    }
+
+    private static final class Replacement implements SerializableOnlyOverRemoting {
+
+        private static final long serialVersionUID = 1;
+
+        private final RemoteOutputStream ros;
+
+        Replacement(BufferedBuildListener cbl) {
+            this.ros = new RemoteOutputStream(new CloseProofOutputStream(cbl.out));
+        }
+
+        private Object readResolve() throws IOException {
+            return new BufferedBuildListener(new DelayBufferedOutputStream(ros));
```

Comment:
OK, will check this.

Reply:
Actually it is not that decoration is taking place on the wrong node (you were perhaps confusing an earlier `Running on remote` message, always printed from master, with the `printed a message on master=false` message being asserted). It is just that with buffering, the remotely printed message is not sent to the build log quickly enough; corrected with

```diff
diff --git a/src/test/java/org/jenkinsci/plugins/workflow/job/console/TaskListenerDecoratorTest.java b/src/test/java/org/jenkinsci/plugins/workflow/job/console/TaskListenerDecoratorTest.java
index 2262458..a53b615 100644
--- a/src/test/java/org/jenkinsci/plugins/workflow/job/console/TaskListenerDecoratorTest.java
+++ b/src/test/java/org/jenkinsci/plugins/workflow/job/console/TaskListenerDecoratorTest.java
@@ -200,6 +200,7 @@ public class TaskListenerDecoratorTest {
             }
             @Override public Void call() throws RuntimeException {
                 listener.getLogger().println("printed a message on master=" + JenkinsJVM.isJenkinsJVM());
+                listener.getLogger().flush();
                 return null;
             }
         }
```

The question is whether this is just a test fix that needs to be made, as I did in `LogStorageTestBase`, or whether this is a real problem and we need to somehow force a flush of any remote streams when a step ends. Let me think this over.

Reply:
And, some `ExportTable.diagnoseInvalidObjectId` warnings. Remoting :man_shrugging:

Reply:
> whether it is practical for the master to reach out to agents at the end of a step and request flushes

I cannot think of any reasonable way to do that. You would need to have the agents `Channel.export` some flushing thunk which the `BufferedBuildListener` would collect so that its `close` could call them all. But at that point the channel might not even be alive, or remote calls could hang, and there could be complicated garbage collection issues here, etc.

I think it is more straightforward to just ask the step, or whatever is working with a `TaskListener` on the remote side, to `flush` at some logical conclusion. This is easily done for `DurableTaskStep` (I will file a PR for it), and for `RemoteLauncher` in core, and for this little case in the test, the change I mentioned. On the flip side, cloud implementations of `TaskListener` need to ensure that `flush` works, something I apparently still need to handle in `CloudWatchSender` in https://github.com/jenkinsci/pipeline-log-fluentd-cloudwatch-plugin/pull/8.

Reply:
Ok, having the remote side flush when it is done with the stream seems like a reasonable solution to me.